### ISSUE TITLE
TCP/IP failed to establish an outgoing connection because the selected local endpoint was recently used to connect to the same remote endpoint

### ISSUE THREAD

**Issue by sepehr1014** [aspnet/KestrelHttpServer#2558]
Using .NET Core 2.1 RC1 on Windows Server 2016, we occasionally get this error and from then on, everything returns HTTP 502.5 error.

`TCP/IP failed to establish an outgoing connection because the selected local endpoint was recently used to connect to the same remote endpoint. This error typically occurs when outgoing connections are opened and closed at a high rate, causing all available local ports to be used and forcing TCP/IP to reuse a local port for an outgoing connection. To minimize the risk of data corruption, the TCP/IP standard requires a minimum time period to elapse between successive connections from a given local endpoint to a given remote endpoint.`

We also observe high memory usage for the worker process (about 1.7GB) when this happens. If we frequently recycle the app pool (say before it hits 3GB of private memory) we never get this error. But we rely on memory cache so that's a painful workaround.

**Comment by Tratcher** [comment]
Context? Where is that error reported?

**Comment by sepehr1014** [comment]
@Tratcher It's not an uncaught exception in my code. From a point forward, all requests yield 502.5 error and that's the only error we found in system event viewer. I'm not exactly sure what causes this error. We're using web sockets if it's related. I also took a dump of the worker process if it helps ...

**Comment by Tratcher** [comment]
@pan-wang

**Comment by pan-wang** [comment]
502.5 means the application failed to start. When did this 502.5 error happen? If the application suddenly returns 502.5, it means the backend process crashed. From the Windows security event logs, you should be able to see some process crash report. ANCM should also log a event in Application log. Could you please check windows event logs?

**Comment by muratg** [comment]
Looks like an ephemeral port exhaustion issue here. Do you have any clue why that may be? Are you able to create a simple repro app?  Anything else interesting going on on the machine?

**Comment by pan-wang** [comment]
@sepehr1014 I need a full picture about this port issue. 1) did client receive 502.5 or other response? if 502.5, ANCM must log an error in windows event log, and have you try the workaround of set a static port using environmental variable 2) if there any windows log showing the crash of the backend process? 3) the high memory usage is just IIS worker process or plus the child (backend) process. By default, the memory usage of child process is included as the usage of IIS worker process. 4) what's the file version of aspnetcore.dll?   
Could you please share the dump of the w3wp when the issue happened. I suspect there could be some kind of connection leak for WebSocket which caused WinHttp failed with the error you saw.

**Comment by sepehr1014** [comment]
@muratg The only thing I can think of is the fact that we use web sockets and there are 200+ active connections at a time. Maybe closed sockets (that didn't send a close frame) are not disconnected properly and are in waiting status. Since when the app pool is recycled every couple of hours, we don't face this issue ...

**Comment by sepehr1014** [comment]
@pan-wang 
1. Yes, it received 502.5 but no other errors are logged in windows events. I haven't tried setting a static port. To dodge this issue, we have set IIS to recycle if private memory reaches 3GB. It seems to work for now but it's far from an ideal solution since we heavily depend on memory cache ...
2. I couldn't find any
3. Just the IIS worker process. I checked the memory usage of w3wp.exe in Task Manager and it had reached 1.7GB (it's usually at 100-200MB)
4. v8.2.1991.0
I just emailed you the dump file.
